<a href="https://colab.research.google.com/github/kuuuun/python_jupyter_notes/blob/main/chapter03_%E8%8A%B1%E5%BC%8F%E7%B4%A2%E5%BC%95.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np

## strides 含义

第一个参数是总内存长度，第二个参数是一个元素占用多少字节

In [ ]:
# 创建一个 3行 4列 的 int32 数组
# 注意：int32 占用 4 个字节 (Bytes)
arr = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12]
], dtype=np.int16)

print("数组形状 (Shape):", arr.shape)
print("数组步进 (Strides):", arr.strides)
print("内存布局标志:\n", arr.flags)

数组形状 (Shape): (3, 4)
数组步进 (Strides): (8, 2)
内存布局标志:
   C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False



In [ ]:
data = np.array([
    [1, 1, 1],
    [2, 2, 2]
])
# Shape = (2, 3) -> (2行, 3列)

print("原始数据:\n", data)

# 场景 A：我要把每一列的数据加起来，变成一个扁的行
# 思考：我要压扁“行”这个维度（垂直方向），所以操作 axis=0
sum_ax0 = data.sum(axis=0)
print(f"\nSum axis=0 (压扁行): {sum_ax0}")
print(f"结果形状: {sum_ax0.shape}") # 形状从 (2,3) 变成了 (3,)

# 场景 B：我要计算每一行的总和，变成一个竖条（或者列表）
# 思考：我要压扁“列”这个维度（水平方向），所以操作 axis=1
sum_ax1 = data.sum(axis=1)
print(f"\nSum axis=1 (压扁列): {sum_ax1}")
print(f"结果形状: {sum_ax1.shape}") # 形状从 (2,3) 变成了 (2,)


原始数据:
 [[1 1 1]
 [2 2 2]]

Sum axis=0 (压扁行): [3 3 3]
结果形状: (3,)

Sum axis=1 (压扁列): [3 6]
结果形状: (2,)


## 降维打击法则

- 不要死记硬背“0是行，1是列”。请记住这个核心口诀：

- “Axis N 是当你剥洋葱时，最先剥掉的那一层。”

- 或者更准确的操作定义：**“当你在某个 Axis 上进行聚合（如求和）时，那个 Axis 就会消失（Collapse）。”**

In [ ]:
data = np.arange(24,dtype=np.int16)

reshape_data=data.reshape(2,3,4)
print(reshape_data)
reshape_data.strides

sum_ax0 = reshape_data.sum(axis=0)
print(f"\nSum axis=0 : {sum_ax0}")
print(f"结果形状: {sum_ax0.shape}") # 形状从 (2,3) 变成了 (3,)
print(f"Strides: {sum_ax0.strides}")

sum_ax1 = reshape_data.sum(axis=1)
print(f"\nSum axis=1 : {sum_ax1}")
print(f"结果形状: {sum_ax1.shape}") # 形状从 (2,3) 变成了 (2,)
print(f"Strides: {sum_ax1.strides}")

sum_ax2 = reshape_data.sum(axis=2)
print(f"\nSum axis=1 : {sum_ax2}")
print(f"结果形状: {sum_ax2.shape}") # 形状从 (2,3) 变成了 (2,)
print(f"Strides: {sum_ax2.strides}")

[[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]

Sum axis=0 : [[12 14 16 18]
 [20 22 24 26]
 [28 30 32 34]]
结果形状: (3, 4)
Strides: (32, 8)

Sum axis=1 : [[12 15 18 21]
 [48 51 54 57]]
结果形状: (2, 4)
Strides: (32, 8)

Sum axis=1 : [[ 6 22 38]
 [54 70 86]]
结果形状: (2, 3)
Strides: (24, 8)


## 什么是视图？

视图就像是透过窗户看原来的数据。你没有复制数据，你只是换了个角度看它。如果你在窗户上涂鸦，原来的风景（数据）看起来也就变了。

In [ ]:
# 创建原始数组
original = np.array([0, 1, 2, 3, 4, 5])

# 创建一个切片：取前3个元素
# 注意：Pythons List 这里会复制，但 NumPy 不会！
my_slice = original[0:3]

print("修改前 Original:", original)
print("修改前 Slice:   ", my_slice)

# 修改切片中的第一个元素
my_slice[0] = 999

print("\n--- 修改切片后 ---")
print("修改后 Slice:   ", my_slice)
print("修改后 Original:", original)
# 惊！Original 的第一个元素也被改成了 999


修改前 Original: [0 1 2 3 4 5]
修改前 Slice:    [0 1 2]

--- 修改切片后 ---
修改后 Slice:    [999   1   2]
修改后 Original: [999   1   2   3   4   5]


In [ ]:
import numpy as np

# 1. 模拟一张 4x4 像素的彩色图片
# 形状: (4, 4, 3) -> (高, 宽, RGB通道)
# 我们用随机整数模拟像素值 (0-255)
img = np.random.randint(0, 256, size=(4, 4, 3), dtype=np.uint8)

print(f"图像形状: {img.shape}")
print(f"数据类型: {img.dtype}")

# 2. 分离通道
# 思考：通道在 axis=2 (最后一个维度)
# img[:, :, 0] 代表所有行、所有列的第 0 个通道 (Red)
red_channel = img[:, :, 0]
green_channel = img[:, :, 1]
blue_channel = img[:, :, 2]

print("\n左上角像素的 RGB 值:", img[0, 0, :])

# 3. 操作：去除红色分量 (滤镜效果)
# 这里我们不想修改原图，所以先 copy 一份
img_no_red = img.copy()
img_no_red[:, :, 0] = 0  # 将 R 通道全部置为 0

print("去除红色后的左上角像素:", img_no_red[0, 0, :])

# 4. 变形：Flatten (展平)
# 机器学习中，经常需要把 3D 图片拉成 1D 向量
# 方法 A: ravel() 返回视图 (更快)
# 方法 B: flatten() 返回副本
flat_view = img.ravel()
flat_copy = img.flatten()

print(f"\n展平后的形状: {flat_view.shape}") # 4*4*3 = 48
print(f"原图是否被修改? {'是' if img.shape == (48,) else '否'}")

# 5. 重塑：Reshape 回去
# 只要元素总数(size)不变，形状可以随意捏
restored_img = flat_view.reshape(4, 4, 3)
print(f"复原后的形状: {restored_img.shape}")

图像形状: (4, 4, 3)
数据类型: uint8

左上角像素的 RGB 值: [ 82 130 209]
去除红色后的左上角像素: [  0 130 209]

展平后的形状: (48,)
原图是否被修改? 否
复原后的形状: (4, 4, 3)


In [ ]:
import numpy as np

# 创建一个 3x4 的矩阵
data = np.array([
    [10, 11, 12, 13],
    [20, 21, 22, 23],
    [30, 31, 32, 33]
])

print("原始数据:\n", data)

# 1. 取第 1 行到第 2 行，第 2 列到第 3 列
# 也就是取中心的那一小块 2x2 矩阵
# 记住：左闭右开区间 [start:end)
sub_region = data[0:2, 1:3]
print("\n中心切片:\n", sub_region)

# 2. 混合使用整数和切片
# 取第 1 行的所有列
row_1 = data[1, :]
print(f"\n第 1 行: {row_1} (Shape: {row_1.shape})")

# 3. 步长切片
# 每隔一行取一行，每隔一列取一列
jump_slice = data[::2, ::2]
print("\n跳跃切片:\n", jump_slice)

原始数据:
 [[10 11 12 13]
 [20 21 22 23]
 [30 31 32 33]]

中心切片:
 [[11 12]
 [21 22]]

第 1 行: [20 21 22 23] (Shape: (4,))

跳跃切片:
 [[10 12]
 [30 32]]


In [ ]:
names = np.array(["Bob", "Joe", "Will", "Bob"])
data = np.random.randn(4, 4)
print(f"data: {data}")
# 提问：哪里是 "Bob"？
mask = (names == "Bob")
print("掩码 (Mask):\n", mask)
# 输出: [True, False, False, True]
type(mask)
bob_data = data[mask]
print("Bob :\n", bob_data)

data: [[-0.6247652  -0.2409731   0.21087085 -0.05340711]
 [-1.31177545  0.51996495 -1.1749854   0.28695553]
 [-0.09561484 -0.03827418  2.03764627 -0.03644275]
 [ 0.76016653 -2.1200994   0.65281845 -0.42095844]]
掩码 (Mask):
 [ True False False  True]
Bob :
 [[-0.6247652  -0.2409731   0.21087085 -0.05340711]
 [ 0.76016653 -2.1200994   0.65281845 -0.42095844]]


### `np.random.default_rng()

`np.random.default_rng()` 是 NumPy 1.17 版本之后推荐使用的随机数生成器创建函数。

简单来说，你可以把它看作是一个“随机数工厂”。调用它会返回一个 Generator 对象，你通过这个对象来生成各种类型的随机数。

相比于旧的 `np.random.rand()` 或 `np.random.randint()`，它更现代、更快、线程安全，而且不会污染全局状态。

In [ ]:
import numpy as np
rng = np.random.default_rng(seed=42)

# 生成一组模拟的温度数据 (包含一些异常值)
temps = rng.integers(-10, 40, size=10)
# 手动制造几个异常噪音
temps[0] = -999
temps[8] = 500

print("原始温度:", temps)

# 1. 查找所有正常范围的数据 (-20 到 50 之间)
# 注意：在 NumPy 中组合条件不能用 'and'/'or'，必须用 '&'/'|'
# 并且每个条件必须加括号！
valid_mask = (temps > -20) & (temps < 50)

print("有效性掩码:", valid_mask)

# 2. 筛选出清洗后的数据
clean_data = temps[valid_mask]
print("清洗后数据:", clean_data)

# 3. 将异常值替换为均值 (数据插补)
# 注意：计算均值时要用 clean_data，否则会受异常值影响
mean_val = int(clean_data.mean())
temps[~valid_mask] = mean_val # '~' 代表逻辑非 (NOT)

print("修复后数据:", temps)

原始温度: [-999   28   22   11   11   32   -6   24  500   -6]
有效性掩码: [False  True  True  True  True  True  True  True False  True]
清洗后数据: [28 22 11 11 32 -6 24 -6]
修复后数据: [14 28 22 11 11 32 -6 24 14 -6]


In [ ]:
rng = np.random.default_rng(12345)
print(rng)
rfloat = rng.random()
print(rfloat)
type(rfloat)

Generator(PCG64)
0.22733602246716966


float

### 这是 Python 程序员转 NumPy 时最大的坑。

Python 的 and 和 or 试图判断整个数组的真假（Truth Value），这在数组含有一个以上元素时会报错。

NumPy 需要对数组的 每个元素 分别做逻辑运算，所以必须使用位运算符 & (AND), | (OR), ~ (NOT)。

In [ ]:
matrix = np.arange(12).reshape(3, 4)
# [[ 0,  1,  2,  3],
#  [ 4,  5,  6,  7],
#  [ 8,  9, 10, 11]]

# 目标：我想要 (0,0) 和 (2,3) 这两个对角线端点
# 写法：分别传入行坐标列表、列坐标列表
rows = [0, 2]
cols = [0, 3]

choice1 = matrix[rows, cols]
print("choice1 :\n", choice1)
# 输出: [ 0 11] -> 也就是 matrix[0,0] 和 matrix[2,3]

# 先取 0，2行，即 matrix[[0,2]], 然后从0，2行中切片，取所有行中的0，3列
# choice2 = matrix[[0,2],...][...,[0,3]]
# choice2 = matrix[[0,2],:][:,[0,3]]
choice2 = matrix[[0,2]][:,[0,3]]
print("choice2 :\n", choice2)

choice3 = matrix[np.ix_([0,2],[0,3])]
print("choice3 :\n", choice3)

choice1 :
 [ 0 11]
choice2 :
 [[ 0  3]
 [ 8 11]]
choice3 :
 [[ 0  3]
 [ 8 11]]


In [ ]:
import numpy as np

# 1. 模拟数据生成
# 列定义: [用户ID, 年龄, 消费金额]
# 生成 10 个用户
data = np.array([
    [101, 25, 200],
    [102, -5, 100],  # 脏数据：年龄 -5
    [103, 30, 0],
    [104, 35, 5000],
    [105, 18, -20],  # 脏数据：金额 -20
    [106, 40, 300],
    [107, 22, 1500],
    [108, 28, 600],
    [109, 50, 450],
    [110, 33, 8000]
])

print(f"原始记录数: {len(data)}")

# 2. 数据清洗 (Boolean Indexing)
# 规则：年龄 > 0 且 金额 >= 0
# 我们关注第 1 列(Age) 和 第 2 列(Amount)
valid_mask = (data[:, 1] > 0) & (data[:, 2] >= 0)

clean_data = data[valid_mask]
print(f"有效记录数: {len(clean_data)}")
print("清洗后数据片段:\n", clean_data[:3])

# 3. 寻找高价值用户
# 计算清洗后的平均消费
avg_spend = clean_data[:, 2].mean()
print(f"\n平均消费: {avg_spend:.2f}")

# 筛选消费 > 平均值的用户
high_value_users = clean_data[clean_data[:, 2] > avg_spend]

print("高价值用户 ID:", high_value_users[:, 0])

# 4. 提取指定 VIP 用户 (Fancy Indexing)
# 假设运营部门给了你一个 VIP 用户 ID 列表，要查他们的消费
vip_ids = [104, 110]

# 注意：这里不能直接用 data[vip_ids]，因为 vip_ids 是 ID 值，不是行索引(Index)
# 我们需要先找到这些 ID 所在的行号
# 使用 np.isin() 这是一个非常实用的布尔判断函数
is_vip = np.isin(clean_data[:, 0], vip_ids)

vip_data = clean_data[is_vip]
print("\nVIP 用户详情:\n", vip_data)

原始记录数: 10
有效记录数: 8
清洗后数据片段:
 [[ 101   25  200]
 [ 103   30    0]
 [ 104   35 5000]]

平均消费: 2006.25
高价值用户 ID: [104 110]

VIP 用户详情:
 [[ 104   35 5000]
 [ 110   33 8000]]


In [ ]:
rng = np.random.default_rng()
data = rng.integers(-10,10,16)

re_data = data.reshape(4,4)
print(f"origin data: \n{re_data}")
re_data[re_data<0] = 0
print(f"modified data: \n{re_data}")

# check base on
print(re_data.base)
print(re_data.base is data)
re_data[0]=-10
print(data)
print(re_data)

origin data: 
[[  9 -10   0   3]
 [ -5   3   1 -10]
 [ -2   9   6   9]
 [  3  -8   2  -1]]
modified data: 
[[9 0 0 3]
 [0 3 1 0]
 [0 9 6 9]
 [3 0 2 0]]
[9 0 0 3 0 3 1 0 0 9 6 9 3 0 2 0]
True
[-10 -10 -10 -10   0   3   1   0   0   9   6   9   3   0   2   0]
[[-10 -10 -10 -10]
 [  0   3   1   0]
 [  0   9   6   9]
 [  3   0   2   0]]


#### 🌟 专业提示 ：

如果你不确定某个操作返回的是 View 还是 Copy，可以检查数组的 .base 属性。如果是 View，.base 会指向原始数组；如果是 Copy，.base 通常是 None 。

In [ ]:
import numpy as np

# 1. 创建原始数组
original = np.array([1, 2, 3, 4])

# 2. 创建视图 (View) - 切片操作通常返回视图
view = original[1:3]
print("视图的 .base 指向原始数组:", view.base is original)  # 输出: True

# 3. 创建副本 (Copy) - copy() 方法返回副本
copy = original.copy()
print("副本的 .base 为 None:", copy.base is None)            # 输出: True

# 4. 验证影响
view[0] = 99  # 修改视图
print("原始数组被视图修改了:", original)  # 输出: [1 99 3 4]

copy[0] = 88 # 修改副本
print("原始数组不受副本影响:", original)  # 输出: [1 99 3 4] (保持不变)

视图的 .base 指向原始数组: True
副本的 .base 为 None: True
原始数组被视图修改了: [ 1 99  3  4]
原始数组不受副本影响: [ 1 99  3  4]
